(DOCSP-23502): Swift SDK IA Updates (#2000)
mongodb/docs-realm: source/sdk/swift/model-data/model-data-device-sync.txt

=======================================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Realm Object Models
-------------------

When you define a Realm object, you derive a class from :swift-sdk:`Object 
<Typealiases.html#/s:10RealmSwift6Objecta>`. The models that you define 
each map to their own collection in MongoDB Atlas. In this example, both
``Dog`` and ``Person`` would map to separate collections in MongoDB Atlas.

.. code-block:: swift

   class Dog: Object {
      @Persisted(primaryKey: true) var _id = ObjectId()
      @Persisted var name = ""
      @Persisted var age = 0
   }

   class Person: Object {
      @Persisted(primaryKey: true) var _id = ObjectId()
      @Persisted var name = ""
   }

JSON Schema
~~~~~~~~~~~

Device Sync automatically converts the Realm Swift data types to their BSON
analogs to store them in Atlas. When the client device syncs the data down
from App Services via Device Sync, the BSON data types become their Swift 
analogs again. If you query for these objects by :ref:`accessing
MongoDB Atlas data directly <ios-mongodb-remote-access>` instead of using

> Device Sync to sync the data, you get the BSON data types. When you remotely 
> access Atlas data, it does not automatically map to the Swift data type equivalents.

Reviewer: ```suggestion
Device Sync to sync the data, you get the BSON data types; MongoDB Data Access does not automatically map to Swift object equivalents.
```
Suggestion: I found the last sentence in this paragraph confusing because "When you remotely access Atlas data" sounds like yet another query method with different limitations. I think this is actually a good place to use a semicolon (or something similar) so the connection between MongoDB Data Access -> BSON data -> automatic mapping becomes clear.

Also suggesting we stick with "Swift object" for Realm object models for consistency with my suggestions in the rest of the paragraph.

      @Persisted var age = 0
   }

   class Person: Object {
      @Persisted(primaryKey: true) var _id = ObjectId()
      @Persisted var name = ""
      // To-many relationship - a person can have many dogs
      @Persisted var dogs: List<Dog>
   }

JSON Schema
```````````

In the App Services schema, we see the new property translates to a field 
``dogs``. The type of this field is an array, the items in the array are 
of type ``objectId``. This is because we defined the primary key on 
our ``Dog`` model as an ``objectId``. This field is an array of the primary 
keys of all of the ``Dog`` objects related to the ``Person`` object.

.. code-block:: json
   :caption: Person with To-Many Relationship to Dog

   {
     "title": "Person",
     "bsonType": "object",
     "required": [
       "_id"
     ],
     "properties": {
       "_id": {
         "bsonType": "objectId"
       },
       "name": {
         "bsonType": "string"
       },
       "dogs": {
         "bsonType": "array",
         "items": {
           "bsonType": "objectId"
         }
       }
     }
   }

Inverse Relationship
~~~~~~~~~~~~~~~~~~~~

Using the objects in the example above, consider a case where the ``Dog``

> object has an inverse relationship to the ``Person`` object. 

Author: I'm torn because the [inverse relationship concept content is on the "Define a Realm Object Model" page](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-23502/sdk/swift/model-data/define-a-realm-object-model/#inverse-relationship). I don't necessarily want to duplicate the concept content on this page; it's more about how Realm objects map to Atlas data via Device Sync. I don't expect people will use _this_ page to define the models; it's more supplemental to the main "Define" page. Do you think it would be sufficient to link back to the concept definitions on the "Define" page?
